keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+
+We can define __Shared Bandwidth__ as a goal for priority band __p__, and
+we have N fixed priority:
+
+- *Defined Shared Bandwidth(p) / DSB(p)*: sum of shared quotas provided by
+buckets at Priority p.
+- *System Total Shared Bandwidth / STSB*: sum\[ DSB(i) over 1..N \] + ESQ
+- *Maximum Usable Shared Bandwidth(p) / MUSB*:
+  - if p == 1: STSB
+  - else: STSB - sum\[ STSB - sum\[MUSB(i) over 1...p-1 \]
+
+Generally, the system's overall shared bandwidth consists of a sum of shared
+quotas defined by all buckets and an auto-tunable extra shared bandwidth. In
+this so-called shared bandwidth, a request w/ higher priority can always be
+executed prior to the requests w/ lower priority. Which is, at worst case,
+when requests from top priority completely occupies the shared bandwidth,
+then requests from lower bands can only be executed under its reserved quota.
+And those "wild" request doesn't match any buckets will be rejected.
+
+##### Overview
+
+Within each priority band, we have a set of buckets matching the priority and
+a WRR request queue. Request that missed reserved quota will be queued into
+the WRR queue w/ the weight of its bucket. Note that reserving quota of
+buckets at priority N can also re-claim requests queued at N.
+
+All requests will be pre-processed before it's actually executed by 3 stages:
+
+0. We find a matched bucket by applying BucketBinding API.
+1. *Reserving*: We check if the bucket's reserved quota. If it's sufficient,
+the request will consume one reserved quota from the bucket and be executed
+immediately.
+2. *Prioritizing*: We classify and enqueue the request into a priority band.
+Requests at higher priority will __always__ be executed prior to requests at lower.
+3. *Distributing*: We poll requests from each priority band's queue by
+[Weighted Round Robin](https://en.wikipedia.org/wiki/Weighted_round_robin)
+algorithm. If the request's matched bucket's weight is higher, its request
+will have more chance to be executed.
+
+Each stage works separately and solves separate goals.
+
+```
+                    +-----------+  enqueued  +--------------+  pending  +--------------+ polled by WRR
+Inbound Request --> | Reserving | -------->  | Prioritizing | --------> | Distributing |  ------->  Executed
+                    +-----------+            +--------------+           +--------------+
+              sufficient |             insufficient |                          |
+          reserved quota |        STSB shared quota |                          |
+                         v                          v                          v
+                 Executed Immediately           Rejected                    Time-out
+```
+
+
+##### API Model and APIServer flags
+
+
+
+```go
+type Bucket struct {
+    ...
+
+    // ReservedQuota defines a max concurrency of requests can run under guarantee w/o competing.
+    ReservedQuota int
+
+    // SharedQuota defines a shared concurrency added to the system's shared bandwidth.
```

Comment:
What does "added to the system's shared bandwidth" mean?

Reply:
EDIT: SharedQuota defines a shared concurrency contributed by the bucket to the quota pool at its priority band

Reply:
I guess I don't understand the model here with buckets donating to a shared pool. What exactly are they donating? I see buckets as consuming from the pool?

Reply:
actually, we're applying a "producer-consumer" pattern here. the pool connects quota producer (which list-watches the models and puts its shared quotas to the pool) and queue drainer (which consumes from the pool and signals the pending request to continue).